### PR TITLE
[SPARK-24971][SQL] remove SupportsDeprecatedScanRow

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This is a follow up of https://github.com/apache/spark/pull/21118 .

In https://github.com/apache/spark/pull/21118 we added `SupportsDeprecatedScanRow`. Ideally data source should produce `InternalRow` instead of `Row` for better performance. We should remove `SupportsDeprecatedScanRow` and encourage data sources to produce `InternalRow`, which is also very easy to build.

## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rdblue @rxin @gatorsmile @gengliangwang

**Comment by jose-torres** [comment]
lgtm

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by rdblue** [comment]
@cloud-fan, I thought it was a requirement to have a committer +1 before merging. Or is this [list of committers](https://spark.apache.org/committers.html) out of date?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala [hunk 1]
```diff
@@ -91,7 +90,7 @@ class RateStreamContinuousReader(options: DataSourceOptions)
         i,
         numPartitions,
         perPartitionRate)
-        .asInstanceOf[InputPartition[Row]]
+        .asInstanceOf[InputPartition[InternalRow]]
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala]: Why is this cast necessary?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamMicroBatchReader.scala [hunk 2]
```diff
@@ -169,7 +170,7 @@ class RateStreamMicroBatchReader(options: DataSourceOptions, checkpointLocation:
     (0 until numPartitions).map { p =>
       new RateStreamMicroBatchInputPartition(
         p, numPartitions, rangeStart, rangeEnd, localStartTimeMs, relativeMsPerValue)
-        : InputPartition[Row]
+        : InputPartition[InternalRow]
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamMicroBatchReader.scala]: Is this needed? Doesn't RateStreamMicroBatchInputPartition implement InputPartition[InternalRow]?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala [hunk 3]
```diff
@@ -121,17 +121,6 @@ class DataSourceV2Suite extends QueryTest with SharedSQLContext {
     }
   }
 
-  test("unsafe row scan implementation") {
-    Seq(classOf[UnsafeRowDataSourceV2], classOf[JavaUnsafeRowDataSourceV2]).foreach { cls =>
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: Why remove unsafe tests?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala [hunk 4]
```diff
@@ -416,12 +405,12 @@ class AdvancedDataSourceV2 extends DataSourceV2 with ReadSupport {
       requiredSchema
     }
 
-    override def planRowInputPartitions(): JList[InputPartition[Row]] = {
-      val lowerBound = filters.collect {
+    override def planInputPartitions(): JList[InputPartition[InternalRow]] = {
+      val lowerBound = filters.collectFirst {
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: Nit: this is an unrelated change.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala [hunk 5]
```diff
@@ -456,57 +445,20 @@ class AdvancedInputPartition(start: Int, end: Int, requiredSchema: StructType)
...
       case "i" => current
       case "j" => -current
     }
-    Row.fromSeq(values)
+    InternalRow.fromSeq(values)
   }
 }
 
 
-class UnsafeRowDataSourceV2 extends DataSourceV2 with ReadSupport {
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: These aren't Row implementations. Why remove them?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala]: I didn't dig into it as the cast was already there. The reason seems to be, `java.util.List` isn't covariant.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamMicroBatchReader.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamMicroBatchReader.scala]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: That's a followup of https://github.com/apache/spark/pull/21118. you removed `SupportsScanUnsafeRow` there and then this test becomes meaningless.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala (hunk 1 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala]: I don't think it's a good idea to leave casts. Can you check to see if this can be avoided? I found in #21118 that many of the casts were unnecessary if variables had declared types and it is much better to avoid explicit casts that work around the type system.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamMicroBatchReader.scala (hunk 2 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamMicroBatchReader.scala]: This is fine since it isn't a cast, but it's generally better to check whether these are still necessary after refactoring.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala (hunk 3 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: Ok.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala (hunk 4 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: Fine by me since it is so small, just wanted to point it out.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala (hunk 5 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: Same answer as above, I'm guessing.

**Comment by rdblue** [comment]
This looks fine other than the possibly unnecessary cast.

**Comment by cloud-fan** [comment]
@rdblue I vaguely remember that, if the PR author himself is a committer, we can merge a PR with one more LGTM from the community and no one objects in several days. I'm sorry if it's not the case.

**Comment by gatorsmile** [comment]
@cloud-fan To be save, let us get one more LGTM from the other committer.

**Comment by gatorsmile** [comment]
retest this please

**Comment by rdblue** [comment]
@cloud-fan, @gatorsmile, I'm fine with that if it's documented somewhere. I wasn't aware of that convention and no one brought it up the last time I pointed out commits without a committer +1.

**Comment by gatorsmile** [comment]
@rdblue I do not think it is documented. Let us be more conservative. Collect LGTM from the committers no matter whether the PR author is a committer or not.

**Comment by gatorsmile** [comment]
It sounds like Github is experiencing a very bad delay. @cloud-fan Could you submit a follow-up PR to address the comments from @rdblue ?

**Comment by cloud-fan** [comment]
addressed in https://github.com/apache/spark/pull/21948#discussion_r206999480

**Comment by rdblue** [comment]
Yeah, I'd say that if it isn't documented then lets go with the usually RTC conventions.
